File: scripts/mem_utils.R
# mem_utils.R — memory instrumentation and budget detection, shared by the R
# scripts that fork worker pools (make_unified_annotation.R,
# calculate_density_batch.R via density_utils.R).
#
# Added after run-000156 (94 Gbp) died in make_unified_annotation with 15 of 55
# mclapply workers killed and no memory figures to size a fix from: a rule's
# benchmark file is only written on success, so a failing run leaves nothing
# behind. These read /proc directly so every run records what it actually used.
#
#   VmRSS  current resident set
#   VmHWM  peak resident set for THIS process. A forked mclapply child inherits
#          the parent's counters, so a child's peak is its whole footprint
#          (shared parent pages + its own) — which is the number to divide the
#          host budget by when sizing the pool.
#   MemAvailable  host headroom, so a run heading for the OOM killer is visible
#          in the log before it dies rather than only in dmesg.
#
# All return NA off Linux; every call site degrades to printing nothing.
#
# NOTE: this file must stay dependency-free (base R only) — it is sourced by
# scripts with very different library stacks.

.mem_field_mb <- function(path, field) {
  if (!file.exists(path)) return(NA_real_)
  ln <- grep(paste0("^", field, ":"), readLines(path, warn = FALSE), value = TRUE)
  if (length(ln) == 0) return(NA_real_)
  suppressWarnings(as.numeric(sub("^[^0-9]*([0-9]+).*$", "\\1", ln[1])) / 1024)
}
mem_rss_mb   <- function() .mem_field_mb("/proc/self/status", "VmRSS")
mem_hwm_mb   <- function() .mem_field_mb("/proc/self/status", "VmHWM")
mem_avail_mb <- function() .mem_field_mb("/proc/meminfo",     "MemAvailable")

mem_str <- function() {
  r <- mem_rss_mb(); h <- mem_hwm_mb(); a <- mem_avail_mb()
  if (is.na(r)) return(NA_character_)
  sprintf("rss=%.1fG peak=%.1fG host_avail=%.1fG", r/1024, h/1024, a/1024)
}

# ── memory budget (scheduler- and cgroup-aware) ──────────────────────────────
# Python mirror: scripts/mem_utils.py — same chain, same source labels, so the
# run_pipeline.py banner, these R gates and TideCluster 1.20's own gates all
# report the same number from the same source. See that file's header for the
# full rationale; the short version is that /proc/meminfo reports the HOST (the
# kernel does not namespace it), so inside a .sif under PBS/Slurm it is the
# node's memory and not the job's limit — TideCluster issue #6, a 128 GB job
# that believed it had 1.6 TB and was OOM-killed hours in.
#
# First hit wins:
#   1. explicit --mem_budget_gb (config max_memory_gb)     x MEMORY_HEADROOM
#   2. AGENT_MEMORY (GB)                                   x MEMORY_HEADROOM
#   3. scheduler env: PBS_RESC_MEM (bytes), SLURM_MEM_PER_NODE (MB),
#      LSB_MAX_MEM_RUSAGE (KB), SLURM_MEM_PER_CPU x SLURM_CPUS_ON_NODE (MB)
#                                                          x MEMORY_HEADROOM
#   4/5. the TIGHTEST of the cgroup headroom (leaf -> root, v1+v2,
#        limit - current usage) and MemAvailable — both are already
#        availability readings, so no headroom factor is applied, and taking
#        the min is right because exceeding either one kills the job
#   6. nothing readable -> NA; callers then skip gating rather than guess
#
# Sources 1-3 name an ALLOCATION (what the job was granted), so MEMORY_HEADROOM
# leaves room for the parent process, the OS cache and tool children.

MEMORY_HEADROOM <- 0.8

.MEM_UNIT_MB <- c(b = 1/1024/1024, k = 1/1024, kb = 1/1024, m = 1, mb = 1,
                  g = 1024, gb = 1024, t = 1024*1024, tb = 1024*1024)

# (variable, unit assumed when the value carries no suffix)
.SCHEDULER_MEM_VARS <- list(c("PBS_RESC_MEM",       "b"),
                            c("SLURM_MEM_PER_NODE", "m"),
                            c("LSB_MAX_MEM_RUSAGE", "k"))
.SCHEDULER_JOB_VARS <- c("PBS_JOBID", "SLURM_JOB_ID", "SLURM_JOBID", "LSB_JOBID")

# Sources that name an allocation (and so carry MEMORY_HEADROOM).
.ALLOCATION_SOURCES <- c("max_memory_gb", "AGENT_MEMORY", "SLURM_MEM_PER_CPU",
                         vapply(.SCHEDULER_MEM_VARS, `[`, character(1), 1))

# Schedulers emit both bare numbers (PBS_RESC_MEM=137438953472) and suffixed
# ones (128gb), so accept either; NA when unparseable.
.parse_mem_to_mb <- function(raw, default_unit) {
  if (is.null(raw) || length(raw) == 0 || is.na(raw) || !nzchar(raw))
    return(NA_real_)
  x <- tolower(trimws(raw))
  m <- regmatches(x, regexec("^([0-9]+(\\.[0-9]+)?)[[:space:]]*([kmgt]?b?)$", x))[[1]]
  if (length(m) == 0) return(NA_real_)
  unit <- if (nzchar(m[4])) m[4] else default_unit
  factor <- unname(.MEM_UNIT_MB[unit])
  if (is.na(factor)) return(NA_real_)
  v <- suppressWarnings(as.numeric(m[2])) * factor
  if (is.na(v) || v <= 0) NA_real_ else v
}

.getenv <- function(name, environ) {
  if (is.null(environ)) Sys.getenv(name, "") else {
    v <- environ[[name]]
    if (is.null(v)) "" else as.character(v)
  }
}

.read_cgroup_num <- function(path) {
  if (!file.exists(path)) return(NA_real_)
  v <- tryCatch(trimws(readLines(path, warn = FALSE)[1]),
                error = function(e) NA_character_, warning = function(w) NA_character_)
  if (is.na(v) || !nzchar(v)) return(NA_real_)
  if (identical(v, "max")) return(Inf)                       # cgroup v2 unlimited
  n <- suppressWarnings(as.numeric(v))
  if (is.na(n)) return(NA_real_)
  # cgroup v1 spells "unlimited" as a huge sentinel (PAGE_COUNTER_MAX).
  if (n > 1e15) return(Inf)
  n
}

# Headroom (MB) inside a single cgroup directory: limit - current usage.
.cgroup_dir_avail_mb <- function(dir, lim_file, cur_file) {
  lim <- .read_cgroup_num(file.path(dir, lim_file))
  if (is.na(lim) || !is.finite(lim)) return(NA_real_)         # absent or unlimited
  cur <- .read_cgroup_num(file.path(dir, cur_file))
  if (is.na(cur) || !is.finite(cur)) cur <- 0
  max(0, (lim - cur) / 1048576)
}

# Every ancestor path of `rel` under `base`, leaf first, then `base` itself.
.cgroup_chain <- function(base, rel) {
  parts <- Filter(nzchar, strsplit(rel, "/", fixed = TRUE)[[1]])
  paths <- character(0)
  for (i in rev(seq_along(parts)))
    paths <- c(paths, file.path(base, paste(parts[seq_len(i)], collapse = "/")))
  c(paths, base)
}

# Under PBS/Slurm the limit is typically set on an ANCESTOR scope (the job
# scope), not on the leaf the process sits in, so walk leaf -> root and take the
# tightest. The bare mount root is probed too: inside a cgroup-namespaced
# container the job's own limits appear there.
cgroup_avail_mb <- function(sysfs_root = "/sys/fs/cgroup",
                            proc_cgroup = "/proc/self/cgroup") {
  lines <- if (!file.exists(proc_cgroup)) character(0) else
    tryCatch(readLines(proc_cgroup, warn = FALSE),
             error = function(e) character(0), warning = function(w) character(0))
  rel_v2 <- ""; rel_v1 <- ""
  v2 <- grep("^0::", lines, value = TRUE)
  if (length(v2) > 0) rel_v2 <- sub("^0::", "", v2[1])
  v1 <- grep(":memory:", lines, value = TRUE)
  if (length(v1) > 0) rel_v1 <- sub("^.*:memory:", "", v1[1])

  cands <- numeric(0)
  for (d in .cgroup_chain(sysfs_root, rel_v2))
    cands <- c(cands, .cgroup_dir_avail_mb(d, "memory.max", "memory.current"))
  for (d in .cgroup_chain(file.path(sysfs_root, "memory"), rel_v1))
    cands <- c(cands, .cgroup_dir_avail_mb(d, "memory.limit_in_bytes",
                                           "memory.usage_in_bytes"))
  cands <- cands[!is.na(cands)]
  if (length(cands) == 0) NA_real_ else min(cands)
}

# Budget in MB. Returns list(mb = <MB or NA>, src = <source label>).
mem_budget_mb <- function(explicit_gb = 0, environ = NULL,
                          sysfs_root = "/sys/fs/cgroup",
                          proc_cgroup = "/proc/self/cgroup",
                          meminfo = "/proc/meminfo") {
  if (!is.na(explicit_gb) && explicit_gb > 0)
    return(list(mb = explicit_gb * 1024 * MEMORY_HEADROOM, src = "max_memory_gb"))

  agent <- .parse_mem_to_mb(.getenv("AGENT_MEMORY", environ), "g")
  if (!is.na(agent)) return(list(mb = agent * MEMORY_HEADROOM, src = "AGENT_MEMORY"))

  for (v in .SCHEDULER_MEM_VARS) {
    mb <- .parse_mem_to_mb(.getenv(v[1], environ), v[2])
    if (!is.na(mb)) return(list(mb = mb * MEMORY_HEADROOM, src = v[1]))
  }
  per_cpu <- .parse_mem_to_mb(.getenv("SLURM_MEM_PER_CPU", environ), "m")
  ncpu <- suppressWarnings(as.integer(.getenv("SLURM_CPUS_ON_NODE", environ)))
  if (!is.na(per_cpu) && !is.na(ncpu) && ncpu > 0)
    return(list(mb = per_cpu * ncpu * MEMORY_HEADROOM, src = "SLURM_MEM_PER_CPU"))

  cands <- c(cgroup = cgroup_avail_mb(sysfs_root, proc_cgroup),
             MemAvailable = .mem_field_mb(meminfo, "MemAvailable"))
  cands <- cands[!is.na(cands)]
  if (length(cands) == 0) return(list(mb = NA_real_, src = "none"))
  list(mb = min(cands), src = names(cands)[which.min(cands)])
}

# One-line summary, identical in wording to the Python mirror's describe_budget.
describe_budget <- function(mb, src) {
  if (is.na(mb))
    return("[mem] no memory budget could be detected (source: none) - worker pools will not be memory-gated")
  detail <- ""
  if (src %in% .ALLOCATION_SOURCES)
    detail <- sprintf(" (%.1f GB allocation x %g)", mb / MEMORY_HEADROOM / 1024,
                      MEMORY_HEADROOM)
  sprintf("[mem] budget %.0f MB from %s%s", mb, src, detail)
}

# Warn when a host-wide budget is in use under a batch scheduler: MemAvailable
# (or nothing at all) while $PBS_JOBID / $SLURM_JOB_ID is set means the gates are
# sizing pools against the NODE's free memory instead of the job's limit — the
# issue-#6 failure, otherwise invisible until the OOM killer arrives.
warn_if_host_budget <- function(src, environ = NULL) {
  if (!src %in% c("MemAvailable", "none")) return(invisible(FALSE))
  for (var in .SCHEDULER_JOB_VARS) {
    value <- .getenv(var, environ)
    if (nzchar(value)) {
      message(sprintf(
        paste0("WARNING: memory budget came from %s, but %s=%s is set. ",
               "MemAvailable is not cgroup-aware, so this is probably the whole ",
               "node's memory, not this job's limit — worker pools may ",
               "over-allocate and be OOM-killed. Pass the job's allocation with ",
               "'run_pipeline.py -m <GB>' (or set max_memory_gb in the config)."),
        if (identical(src, "MemAvailable")) "/proc/meminfo MemAvailable" else "nowhere",
        var, value))
      return(invisible(TRUE))
    }
  }
  invisible(FALSE)
}

File: scripts/mem_utils.py
#!/usr/bin/env python3
"""Memory-budget resolution, shared by the Python side of the pipeline.

The Python mirror of ``scripts/mem_utils.R``. Both resolve the same chain and
return the same source labels, so every layer of a run — the ``run_pipeline.py``
banner, the R worker-pool gates, and TideCluster's own gates — reports the same
number from the same source.

WHY THIS EXISTS
---------------
CARP normally runs from a ``.sif`` under PBS/Slurm. ``/proc/meminfo``
``MemAvailable`` reports the **host**, not the job — the kernel does not
namespace it — and the cgroup limit that will actually kill the job is usually
set on an ANCESTOR scope (the job scope), which may not be reachable from inside
the container at all: controllers not delegated, ``/sys/fs/cgroup`` not mounted,
or a scheduler that enforces by polling rather than by cgroup. A pool sized from
host memory then over-allocates by an order of magnitude and the run is
OOM-killed hours in, with no earlier warning (TideCluster issue #6: a 128 GB PBS
job believed it had 1.6 TB).

The chain below is deliberately identical to TideCluster 1.20.0's
``tc_utils.memory_budget_mb()``. First hit wins:

  1. explicit budget      -- config ``max_memory_gb`` / ``--mem_budget_gb``
  2. ``AGENT_MEMORY``     -- GB, sandbox override
  3. scheduler env        -- ``PBS_RESC_MEM`` (bytes), ``SLURM_MEM_PER_NODE`` (MB),
                             ``LSB_MAX_MEM_RUSAGE`` (KB),
                             ``SLURM_MEM_PER_CPU`` x ``SLURM_CPUS_ON_NODE`` (MB).
                             These survive the container boundary, which is what
                             makes them the reliable signal under a scheduler.
  4/5. the tightest measured reading -- cgroup headroom (v2 ``memory.max`` /
                             v1 ``memory.limit_in_bytes``, minus current usage)
                             and ``/proc/meminfo`` ``MemAvailable``
  6. nothing readable     -- ``(None, "none")``; callers skip gating rather than guess

Sources 1-3 are ALLOCATIONS (what the job was granted), so ``MEMORY_HEADROOM``
is applied: a pool sized to exactly the job limit leaves nothing for the parent
process, the OS cache or tool children. Sources 4-5 are already
measured-availability numbers (cgroup ``limit - current``, ``MemAvailable``) and
pass through unchanged.

Three deliberate differences from upstream, all retained from mem_utils.R: the
cgroup rung reports ``limit - current`` rather than the raw limit; it walks
leaf -> root so a limit set on an ancestor job scope is picked up; and the two
measured rungs are combined as ``min()`` rather than first-hit, because
exceeding either one kills the job.

Dependency-free (stdlib only): ``run_pipeline.py`` imports it before any conda
environment exists. Every filesystem root is injectable so the tests can build
fake ``/sys/fs/cgroup`` and ``/proc`` trees.
"""
from __future__ import annotations

import os
import re
import sys

# Fraction of an ALLOCATION a worker pool may plan to use. Matches
# TideCluster 1.20.0's MEMORY_HEADROOM, so both agree on the same job.
MEMORY_HEADROOM = 0.8

# cgroup v1 spells "unlimited" as a huge sentinel (PAGE_COUNTER_MAX).
_CGROUP_UNLIMITED = 1e15

_MEM_UNIT_MB = {
    "b": 1.0 / 1024 / 1024,
    "k": 1.0 / 1024, "kb": 1.0 / 1024,
    "m": 1.0, "mb": 1.0,
    "g": 1024.0, "gb": 1024.0,
    "t": 1024.0 * 1024, "tb": 1024.0 * 1024,
}

# (variable, unit when the value carries no suffix)
_SCHEDULER_MEM_VARS = (
    ("PBS_RESC_MEM", "b"),        # PBS/Torque: bytes
    ("SLURM_MEM_PER_NODE", "m"),  # Slurm: MB
    ("LSB_MAX_MEM_RUSAGE", "k"),  # LSF: KB
)

_SCHEDULER_JOB_VARS = ("PBS_JOBID", "SLURM_JOB_ID", "SLURM_JOBID", "LSB_JOBID")

_MEM_RE = re.compile(r"(\d+(?:\.\d+)?)\s*([kmgt]?b?)")


def _parse_mem_to_mb(raw, default_unit: str):
    """Parse a scheduler memory string to MB. ``None`` if unparseable.

    Accepts a bare number in ``default_unit`` (``PBS_RESC_MEM=137438953472``)
    or a suffixed one (``128gb``, ``2048mb``) — schedulers emit both.
    """
    if raw is None:
        return None
    m = _MEM_RE.fullmatch(str(raw).strip().lower())
    if not m:
        return None
    factor = _MEM_UNIT_MB.get(m.group(2) or default_unit)
    if factor is None:
        return None
    value = float(m.group(1)) * factor
    return value if value > 0 else None


def _read_cgroup_num(path: str):
    """Numeric content of a cgroup control file.

    Returns ``None`` when absent/unreadable and ``float("inf")`` for the two
    spellings of "unlimited" (v2 ``max``, v1 sentinel).
    """
    try:
        with open(path) as fh:
            raw = fh.read().strip()
    except (OSError, ValueError):
        return None
    if not raw:
        return None
    if raw == "max":
        return float("inf")
    try:
        value = float(raw)
    except ValueError:
        return None
    return float("inf") if value > _CGROUP_UNLIMITED else value


def _cgroup_dir_avail_mb(directory: str, lim_file: str, cur_file: str):
    """Headroom (MB) inside one cgroup directory: ``limit - current usage``."""
    limit = _read_cgroup_num(os.path.join(directory, lim_file))
    if limit is None or limit == float("inf"):
        return None                                  # absent or unlimited
    current = _read_cgroup_num(os.path.join(directory, cur_file))
    if current is None or current == float("inf"):
        current = 0.0
    return max(0.0, (limit - current) / 1048576)


def _chain(base: str, rel: str) -> list:
    """``base`` plus every ancestor path of ``rel`` under it, leaf first."""
    paths, path = [], rel
    while path and path != "/":
        paths.append(base + path)
        path = os.path.dirname(path)
    paths.append(base)
    return paths


def cgroup_avail_mb(sysfs_root: str = "/sys/fs/cgroup",
                    proc_cgroup: str = "/proc/self/cgroup"):
    """Tightest effective cgroup headroom in MB, or ``None`` if there is none.

    Walks leaf -> root because the limit that kills a batch job is typically set
    on an ancestor scope, not on the leaf the process sits in. The bare mount
    root is probed too: inside a cgroup-namespaced container the job's own
    limits appear there.
    """
    rel_v2, rel_v1 = "", ""
    try:
        with open(proc_cgroup) as fh:
            for line in fh:
                parts = line.strip().split(":", 2)
                if len(parts) != 3:
                    continue
                if parts[0] == "0" and not rel_v2:                     # v2 unified
                    rel_v2 = parts[2]
                elif "memory" in parts[1].split(",") and not rel_v1:   # v1
                    rel_v1 = parts[2]
    except OSError:
        pass

    candidates = []
    for path in _chain(sysfs_root, rel_v2):
        mb = _cgroup_dir_avail_mb(path, "memory.max", "memory.current")
        if mb is not None:
            candidates.append(mb)
    for path in _chain(os.path.join(sysfs_root, "memory"), rel_v1):
        mb = _cgroup_dir_avail_mb(path, "memory.limit_in_bytes",
                                  "memory.usage_in_bytes")
        if mb is not None:
            candidates.append(mb)
    return min(candidates) if candidates else None


def mem_available_mb(meminfo: str = "/proc/meminfo"):
    """``MemAvailable`` in MB — the HOST's, inside a container. ``None`` off Linux."""
    try:
        with open(meminfo) as fh:
            for line in fh:
                if line.startswith("MemAvailable:"):
                    return int(line.split()[1]) / 1024
    except (OSError, ValueError, IndexError):
        pass
    return None


def memory_budget_mb(explicit_gb=None, environ=None,
                     sysfs_root: str = "/sys/fs/cgroup",
                     proc_cgroup: str = "/proc/self/cgroup",
                     meminfo: str = "/proc/meminfo"):
    """Resolve the memory budget. Returns ``(budget_mb, source)``.

    ``source`` is one of ``max_memory_gb``, ``AGENT_MEMORY``, a scheduler
    variable name, ``cgroup``, ``MemAvailable`` or ``none`` — see the module
    docstring for the full chain. ``budget_mb`` is ``None`` only for ``none``.
    """
    env = os.environ if environ is None else environ

    if explicit_gb:
        return float(explicit_gb) * 1024 * MEMORY_HEADROOM, "max_memory_gb"

    agent = _parse_mem_to_mb(env.get("AGENT_MEMORY"), "g")
    if agent:
        return agent * MEMORY_HEADROOM, "AGENT_MEMORY"

    for var, unit in _SCHEDULER_MEM_VARS:
        mb = _parse_mem_to_mb(env.get(var), unit)
        if mb:
            return mb * MEMORY_HEADROOM, var

    per_cpu = _parse_mem_to_mb(env.get("SLURM_MEM_PER_CPU"), "m")
    ncpu = (env.get("SLURM_CPUS_ON_NODE") or "").strip()
    if per_cpu and ncpu.isdigit() and int(ncpu) > 0:
        return per_cpu * int(ncpu) * MEMORY_HEADROOM, "SLURM_MEM_PER_CPU"

    # Rungs 4-5 are both measured-availability readings, so take the TIGHTEST
    # rather than the first hit: exceeding either one kills the job. The label
    # names whichever won (cgroup on a tie).
    measured = [(mb, src) for mb, src in (
        (cgroup_avail_mb(sysfs_root=sysfs_root, proc_cgroup=proc_cgroup), "cgroup"),
        (mem_available_mb(meminfo=meminfo), "MemAvailable")) if mb]
    if measured:
        return min(measured, key=lambda pair: pair[0])

    return None, "none"


def scheduler_job_id(environ=None):
    """``(variable, value)`` of the first batch-scheduler job id set, else ``None``."""
    env = os.environ if environ is None else environ
    for var in _SCHEDULER_JOB_VARS:
        value = (env.get(var) or "").strip()
        if value:
            return var, value
    return None


# Sources that name an ALLOCATION (and therefore carry MEMORY_HEADROOM), as
# opposed to a measured-availability reading (cgroup headroom, MemAvailable).
_ALLOCATION_SOURCES = frozenset(
    ("max_memory_gb", "AGENT_MEMORY", "SLURM_MEM_PER_CPU")
    + tuple(var for var, _unit in _SCHEDULER_MEM_VARS))


def describe_budget(budget_mb, source) -> str:
    """The one-line ``[mem] …`` summary printed at startup and by the gates."""
    if budget_mb is None:
        return ("[mem] no memory budget could be detected (source: none) - "
                "worker pools will not be memory-gated")
    detail = ""
    if source in _ALLOCATION_SOURCES:
        detail = (F" ({budget_mb / MEMORY_HEADROOM / 1024:.1f} GB allocation"
                  F" x {MEMORY_HEADROOM:g})")
    return F"[mem] budget {budget_mb:.0f} MB from {source}{detail}"


def warn_if_host_budget(source, environ=None, stream=None) -> bool:
    """Warn when a host-wide budget is in use under a batch scheduler.

    ``MemAvailable`` (or nothing at all) while ``$PBS_JOBID`` / ``$SLURM_JOB_ID``
    is set means the memory gates are sizing pools against the NODE's free
    memory instead of the job's limit — the issue-#6 failure, which otherwise
    only becomes visible when the OOM killer arrives. Returns whether it warned.
    """
    if source not in ("MemAvailable", "none"):
        return False
    job = scheduler_job_id(environ)
    if job is None:
        return False
    var, value = job
    print(F"WARNING: memory budget came from "
          F"{'/proc/meminfo MemAvailable' if source == 'MemAvailable' else 'nowhere'}, "
          F"but {var}={value} is set. MemAvailable is not cgroup-aware, so this is "
          F"probably the whole node's memory, not this job's limit — worker pools "
          F"may over-allocate and be OOM-killed. Pass the job's allocation with "
          F"'run_pipeline.py -m <GB>' (or set max_memory_gb in the config).",
          file=sys.stderr if stream is None else stream)
    return True


def main() -> int:
    """Report what this machine/job resolves to — a debugging aid.

    Useful inside a container to see what the pipeline will believe:
    ``singularity exec image.sif /opt/pipeline/scripts/mem_utils.py``
    """
    import argparse
    parser = argparse.ArgumentParser(description=__doc__.split("\n")[0])
    parser.add_argument("--max_memory", "--max-memory", type=float, default=None,
                        metavar="GB", help="explicit allocation to resolve against")
    args = parser.parse_args()
    budget_mb, source = memory_budget_mb(args.max_memory)
    print(describe_budget(budget_mb, source))
    warn_if_host_budget(source)
    job = scheduler_job_id()
    print(F"[mem] scheduler job: {job[0]}={job[1]}" if job
          else "[mem] no batch-scheduler job id in the environment")
    return 0


if __name__ == "__main__":
    sys.exit(main())
